### DISCUSSION TITLE
Modoboa Installation fails on Centos 9 Stream

### DISCUSSION THREAD

**Discussion by tomas-kucera** [modoboa/modoboa#2721] [Installer]
# Impacted versions

* OS Type: Centos
* OS Version: 9 Stream
* Database Type: PostgreSQL
* Database version: 13.7
* Modoboa: 2.0.3
* installer used: Yes
* Webserver: Nginx

# Steps to reproduce
- clean Centos 9 Stream installation
- run Modoboa installation according to the documentation
```
> git clone https://github.com/modoboa/modoboa-installer
> cd modoboa-installer
> sudo ./run.py <your domain> --debug
```
# Current behavior
1. Firstly I got error complaining about rrdtool-devel package - solved it by using EPEL repo
https://docs.fedoraproject.org/en-US/epel/#_el9
```
dnf config-manager --set-enabled crb
dnf install epel-release epel-next-release
```

2. Now I am getting error complaining about `uwsgi-plugin-python36`.
```
Error: Unable to find a match: uwsgi-plugin-python36
Failed to install dependencies
```

I could only find and install `uwsgi-plugin-python3`, but not explicitly `uwsgi-plugin-python36`. The whole installation fails on this.

# Expected behavior

Provide some guidance on prerequisites before the installation. I am sure it is difficult with all the combinations.

**Comment by tomas-kucera** [comment]
Managed to get it solved:

- enable EPEL repos:
```
dnf config-manager --set-enabled crb
dnf install epel-release epel-next-release
```

- install the uwsgi:
```
dnf install uwsgi uwsgi-plugin-python3
```

- go to `modoboa-installer/modoboa_installer/scripts/uwsgi.py` and find line `"rpm": ["uwsgi", "uwsgi-plugin-python36"],` and just delete the last `6`

- now (re)run the `run.py`

Unfortunately it does not install on Centos 9 Stream anyway as further on it starts complaining about missing /etc/yum.repos.d/CentOS-Base.repo  from modoboa-installer/modoboa_installer/scripts/postfix.py line 39!

But this is another issue.

It would be great if someone checked and validated this solution and potentially incorporated it into the installation script, so it would recognize the Centos 9 and did what is needed or just gave instructions.

**Comment by Spitfireap** [comment]
Hi, I've made some work on it ([available here](https://github.com/modoboa/modoboa-installer/pull/441)). You need to disable `SE Linux` though. Keep in mind that Centos 9 Stream is not yet officially supported :p.

**Comment by tomas-kucera** [comment]
I know, I understand and have no trouble to do some research and some manual work. ;)

I do not have access to the PR. If there is anything, I can help with testing, I can.

Currently I am playing with Centos 9 and Modoboa. Would like to make it a production server. As of now I can be easily destroying and re-installing the VM. So ... great time to test.

I do not know how far you got with the installation and configuration.

**Comment by Spitfireap** [comment]
> I do not have access to the PR

You should be able to access it : `git clone https://github.com/Spitfireap/modoboa-installer.git && cd modoboa-installer && git checkout centos9-support`. I haven't played with it since October but I got it working (minus the SE linux thing)

**Comment by tomas-kucera** [comment]
OK, I can access it, although in the issue it said that to be able to access the PR I need write privileges.

Already testing it and with clean install of CentOS 9 it went through smoothly! 👍

Except this error at the end ...
```
Redirecting to /bin/systemctl restart cron.service
Failed to restart cron.service: Unit cron.service not found.
Congratulations! You can enjoy Modoboa at https://mail.example.com (admin:password)
```

Could you, please, elaborate more on the SELinux? Why and when it has to be disabled?

I am now getting `Bad request` error but that goes down to the domain name. The installer has one really bad (to my opinion) feature and that is the automatic modification of the domain by prefixing it with mail.xxxx! And seems like it is then spread across many places. Any way to enforce the domain I want as I do not want it to start with the mail? I know there is some logic behind it but... 😄 

Thanks a lot in advance!

**Comment by Spitfireap** [comment]
Feel free to open a PR on the branch if you want. I'll check the cron stuff when I have time. 

The bad request is likely to be due to the SE Linux set to enforce. Try to set it to permissive (`sudo setenforce 0` if I remember correctly). Basically, SE Linux prevents Nginx to access modoboa's or uwsgi files because they are not in the same domain policy (I do not have plentiful of experience with this security). If you want more info, you can take a look [here](https://www.nginx.com/blog/using-nginx-plus-with-selinux/)

As for the prefixing, I agree it's not optimal (I'm working on a QoL improvement in that regard). If you want to get rid of it : `./run.py --stop-after-configfile-check abc` then edit the line `hostname = mail.%(domain)s` on `installer.cfg` then proceed with the installation :). The `mail` bit is to get something unified, and is perhaps required with an RFC. The installer will modify the host to reflect the selected hostname (to prevent being rejected by pretty much any other mail server ;) ).

**Comment by tomas-kucera** [comment]
Hi there, hope you all had great Holidays.

I finally got back to "playing" with modoboa and CentOS 9. It looks like there was some change as the centos9 branch is now failing (which did not before).

During the spamassassin installation it fails with this message:
```
Traceback (most recent call last):
  File "/root/modoboa-installer/./run.py", line 256, in <module>
    main(sys.argv[1:])
  File "/root/modoboa-installer/./run.py", line 240, in main
    scripts.install(appname, config, args.upgrade, args.restore)
  File "/root/modoboa-installer/modoboa_installer/scripts/__init__.py", line 29, in install
    getattr(script, appname.capitalize())(config, upgrade, archive_path).run()
  File "/root/modoboa-installer/modoboa_installer/scripts/base.py", line 191, in run
    self.post_run()
  File "/root/modoboa-installer/modoboa_installer/scripts/amavis.py", line 95, in post_run
    install("spamassassin", self.config, self.upgrade, self.archive_path)
  File "/root/modoboa-installer/modoboa_installer/scripts/__init__.py", line 29, in install
    getattr(script, appname.capitalize())(config, upgrade, archive_path).run()
  File "/root/modoboa-installer/modoboa_installer/scripts/base.py", line 191, in run
    self.post_run()
  File "/root/modoboa-installer/modoboa_installer/scripts/spamassassin.py", line 62, in post_run
    install("razor", self.config, self.upgrade)
TypeError: install() missing 1 required positional argument: 'archive_path'
```

So, I modified the line 62 to read (taking postfix.py as an example):
`install("razor", self.config, self.upgrade, self.archive_path)`
instead of the original:
`install("razor", self.config, self.upgrade)`

After that I ran into two more installation issues:

**Issue 1:**
Redirecting to /bin/systemctl start opendkim.service
Job for opendkim.service failed because the control process exited with error code.
See "systemctl status opendkim.service" and "journalctl -xeu opendkim.service" for details.

Details:
Dec 26 12:51:02 fail.flanelo.cz opendkim[18697]: opendkim: /etc/opendkim.conf: configuration error at line 94: unrecognized parameter
Dec 26 12:51:02 fail.flanelo.cz systemd[1]: opendkim.service: Control process exited, code=exited, status=78/CONFIG

The problem is that the installation scripts adds `SOCKET="inet:12345@localhost"` at the bottom of the `/etc/opendkim.conf` but it is already there on line 42 `Socket inet:12345@localhost`!

Fix: In the file `./scripts/opendkim.py` delete / comment out lines 95-99 that update the config file.
```
with open(params_file, "a") as f:
  f.write('\n'.join([
    "",
    'SOCKET="inet:12345@localhost"',
  ]))
```

**Issue 2:**
The `run.py` script tries to restart cron service with `system.restart_service("cron")` but the service is crond and thus on line 241 of the `run.py` script just change it to `system.restart_service("crond")`.

After these changes I was able to install and run Modoboa on CentOS 9 without any errors and issues!

I am sure that the "fixes" need more attention, such as checking for the OS type / version / existence of the config entries, etc., not just the handmade hardcoded fixes I did! But I will "generously" leave it to the professionals! ;)

@Spitfireap  - Do you want to take a fix these three things (spamassassin, opendkim, crond). I would the gladly contribute an installation guide of Modoboa on CentOS 9 Stream. I already have it for myself anyway (including the fixes).

**Comment by Spitfireap** [comment]
Thanks for the help, I've committed your fixes if you want to give it a try. You didn't have any issue with `SE Linux` on your side ?

**Comment by tomas-kucera** [comment]
> Thanks for the help, I've committed your fixes if you want to give it a try. You didn't have any issue with `SE Linux` on your side ?

I took your advice to disable the SE Linux. ;) I am absolutely not a SE Linux Pro.

**Comment by tomas-kucera** [comment]
BTW, I am currently doing some more tests and configs and running into issues. Some of them might be because I screwed something up and some may be more generic.

What would be the best platform to discuss such? Discord? I am starting to like Modoboa more and more and would like to help improving if I am able either by testing, analyzing, even fixing or leading to fixes. ;)

**Comment by Spitfireap** [comment]
You can join [modoboa's discord](https://discord.gg/WuQ3v3PXGR) and send me a PM :)
